File: test_algorithms/python/radix_sort.py
# Radix Sort (LSD) Implementations for AlgoDisplay Python Test Suite

async def _count_sort_by_digit(arr, n, exp):
    out = [0] * n
    cnt = [0] * 10

    for i in range(n):
        cnt[(arr[i] // exp) % 10] += 1

    for i in range(1, 10):
        cnt[i] += cnt[i - 1]

    for i in range(n - 1, -1, -1):
        d = (arr[i] // exp) % 10
        out[cnt[d] - 1] = arr[i]
        cnt[d] -= 1
        await compare(i, i)

    for i in range(n):
        arr[i] = out[i]

    await render_array(arr)


async def radix_sort(arr):
    """
    Radix Sort (LSD - Least Significant Digit)
    """
    if len(arr) <= 1:
        return arr

    n = len(arr)
    m = max(arr)
    exp = 1

    while m // exp > 0:
        await _count_sort_by_digit(arr, n, exp)
        exp *= 10

    return arr

File: test_algorithms/javascript/radix-sort.js
// Radix Sort (LSD) Implementations for AlgoDisplay Test Suite

/**
 * Radix Sort (LSD - Least Significant Digit)
 */
async function radixSort(arr) {
    if (arr.length <= 1) return arr;

    const countSortByDigit = async (arr, n, exp) => {
        const out = new Array(n).fill(0);
        const count = new Array(10).fill(0);

        for (let i = 0; i < n; i++) {
            count[Math.floor(arr[i] / exp) % 10]++;
        }

        for (let i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        for (let i = n - 1; i >= 0; i--) {
            const d = Math.floor(arr[i] / exp) % 10;
            out[count[d] - 1] = arr[i];
            count[d]--;
            await compare(i, i);
        }

        for (let i = 0; i < n; i++) {
            arr[i] = out[i];
        }
        await renderArray(arr);
    };

    const n = arr.length;
    const maxVal = Math.max(...arr);

    for (let exp = 1; Math.floor(maxVal / exp) > 0; exp *= 10) {
        await countSortByDigit(arr, n, exp);
    }

    return arr;
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        radixSort
    };
}
